README.md
# Exile
Created by Dokimi  
for `Minetest 5.4`

CAUTION: _Some people may find parts of this game difficult or disturbing._

## Installation
Installation from the Minetest `ContentDB` is preferred; just add _Exile_ from the in-game menu &gg; `Content` &gg; `Browse online content` &gg; search in the textbox &gg; click on the `Install` plus sign.

If extracting manually, _Exile_ should be placed in the `games` folder of your user’s Minetest settings, like so:

	Minetest > games > exile > mods/menu/.../etc.

Where `Minetest` is `~/.minetest` in most Linux, and macOS configurations, `~/.var/app/net.minetest.Minetest/.minetest` in Linux Flatpak, and `%USERHOME%\Roaming\Minetest` in most Windows versions.

Since `0.2.3`, _Exile_ requires `naturalslopeslib` (`nsl`). If you’re installing the source from `git`, you’ll need to install `nsl` as well. Extract it into the `./Exile/mods/` folder or use `git pull --recurse submodules` if you have cloned the repository.

## Gameplay
Challenging, at times brutal, wilderness survival with simple technology.
Use your wits to find food, water, and shelter before succumbing to the elements, while exploring the mysterious world, and developing your capacities to endure your exile.

### Features
*Player health effects* — Hypothermia, exhaustion, disease, …  
*Dynamic natural world* — Seasonal weather, erosion, water flows through soil, …  
*Plausible building materials* — Make shelters from the rain, kilns, smelters, …  

## World Settings
Valleys is the standard `mapgen` for _Exile_. `Carpathian` is also supported, for a somewhat more difficult and slower-paced game.

`Flat` mostly works, but `Merki`, and therefore the `Glow Paint` and `Herbal Medicine` made from it, will be unavailable. Enable it in `game.conf` if you’d like to try it anyway.

## Gameplay Guide
Check out `doc > walkthrough` for a more detailed guide.

Many different strategies might work, and part of the fun is figuring out what does, and catastrophically does not, work. &smile; &frown;

Here are some early steps you might pass though:
1. **Craft**   basic tools. **Find** a suitable camp site quickly before you get tired.
1. **Build**   a bed to rest in, plus shelter, and maybe fires for warmth.
1. **Weave**   rudimentary clothing and blankets to begin improving your rest rate.
1. **Make**    a kiln, and fire pots. **Harvest** wild foods from trees while you wait.
1. **Collect** water with pots, and build up a freshwater supply, into a cistern if needed.
1. **Farm**    food plants, herbs, and fibrous canes to build up supplies.
1. **Explore** and gather mineral resources for more advanced tools.

You start in spring. Maybe some pangs of winter linger, so beware the night.
Enjoy the rains while you can. Soon the hot, dry weather of summer will arrive.
Running out of water, or becoming exhausted from the heat, is a real risk.
After that will come the sub-zero conditions of winter.
Starvation and freezing are hard to avoid without preparation.

### Some Tips and Tricks
- **Beds**           are important: If you’re exhausted you get hypothermia/heat stroke.
                     Get under shelter in a nice temperature.
- **Time**:          Do your crafting, organizing, planning while you rest.
- **Weather**:       Extremes sap your energy.
                     Like real life, walking around in a snowstorm is a bad idea!
- **Temperature**:   Build a shelter, with a fire place or lots of torches.
                     You can also go underground; caves sufficiently deep are safe from weather.
- **Water**:         You can drink cave drips (click them).
                     Water pots collect rain water.
                     Some plants, and food quench your thirst.
                     If you’re desperate, you can try to melt ice, 
                     or dig a seepage pit in wet ground and wait for water to flow into it.
- **Food**:          Eat things and see if you live!
                     You can catch animals with clubs (right-click while wielding).
- **Clothing**:      Simple grass clothing can be woven quickly, and can be a lifesaver.
                     Better clothes can be made by soaking bundles of the right kind of plants.
- **Farming**:       Digging tools also can till soil.
                     Punching depleted agricultural soils with fertilizers will restore them.
- **Illness**:       Keep an eye on “health effects.”
                     You may have eaten something bad, or have a terrible disease.
- **Drugs**:         Some plants have useful medicinal effects.
                     Be careful not to overdose, however.
- **Spelunking**:    Go deep enough underground and you might find something….
- **Woodcutting**:   Hardwood trees are more difficult to cut than soft wood.
                     You will need better tools for those.
- **Climbing**:      Build stairs and shelters around your base.
                     This will save you energy and protect you from extremes.
                     You can use sticks to build a ladder, or a pole to shimmy up, 
                     to get to high places, or descend if you’re careful.
- **Environment**:   Not every step in crafting can be done at a work station.
                     Some things need to be fired, or soaked in water, etc.
- **Industry**:      Build ovens, kilns, furnaces the same as you would for real.
                     Start a fire with access to air, and a sealed chamber that gets heated up.
- **Fires**:         Blocks are hotter than slabs. Charcoal is hotter than wood.
                     Fires can be temporarily extinguished by punching while them holding sediment.
- **Charcoal**:      Make it as you would for real — make a wood fire sealed up to limit airflow.
- **Glassmaking**:   Sand and wood-ash can be made into green glass.
                     The ash must be soaked, dried, and roasted to make clear glass — 
                     Glass can be melted onto iron trays to make panes for real windows!
- **Iron smelting**: This is hard. It needs plenty of charcoal, 
                     and a space below the iron mixture for slag to drain out.

## Settings for Multiplayer
The variable `time_speed` defaults to `72`, and at this rate a player who logs on at 8:00 AM every day will get a change in season every 2.5 days.

Changing speed to `60` will make days last 24 minutes and a new season every real-world day, but he will see the seasons in reverse. At `time_speed` of `96`, days last only 15 minutes, but the player might see spring (year 1) on day 1, summer (year 2) on day 2, etc.

Set `exile_hud_update` to `1.0` second for multiplayer servers on the Internet; a LAN server can probably handle `0.2` seconds.

## Mods for Multiplayer
- [Wield3d](https://github.com/stujones11/wield3d) is recommended.
- [`Alternode`](https://github.com/AntumMT/mod-alternode) was used to add the “infotext” popups in the spawn shelter on the [Land of Catastrophe](https://exile.planetofnix.com/wiki/pmwiki.php?n=Loc.LandOfCatastrophe) server.

## Development
_Exile_ is open-source software — that means the game is as good as you choose to make it. It also means development can be erratic and haphazard at times, so be patient!

_Exile_ is currently in “Alpha,” therefore you can expect that there may be bugs, missing features, performance issues, and perhaps compatibility-breaking updates.
Despite this, _Exile_ does have enough features to be a playable game and should be stable and mostly bug-free.

See the [GitHub repository](https://github.com/jeremyshannon/Exile/) for known bugs, and to report new ones.

## Credits
Gratitude is due to all those whose mods have been adapted for use in _Exile_ (see `./mod/` folders for details).

Thanks also to all who have given feedback, fixes, etc.  
A full, up-to-date list of contributors can be found on the GitHub repository, under the `Insights` tab.


CHANGELOG.md
# Changelog:

## [2022-08-05][2022-08-05__0.3.8d] — 0.3.8d and 0.2.11d
  - Minor bugfixes

## [2022-08-05][2022-08-05__0.3.8c] — 0.3.8c and 0.2.11c
  - enabled shadows when running on MT 5.6.0+
  - Pots fill with rainwater even if the player is away
  - Many, many bug fixes

## [2022-07-08][2022-07-08__0.3.8b] — 0.3.8b and 0.2.11b
  - Cheaper rammed earth
  - Improved Sarkamos life cycle and behavior
  - Cooking pot improvements
  - Fixed wooden & iron chests' trashcan icon, dig & replace to reload
  - Misc bugfixes

## [2022-06-25][2022-06-25__0.3.8a] — 0.3.8a and 0.2.11a
  - Reverted betterfall mod due to problems on some systems

## [2022-06-24][2022-06-24__0.3.8] — 0.3.8 and 0.2.11
  - New tech rebalance, multiple ways up the tech tree
  - Massive rings left scattered on the surface by the Ancients
  - Throwable spears for hunting and fighting
  - Improved canoe and airboat performance; the airboat strafes with AUX1
  - Fire resistance for tree trunks and other nodes
  - Ability to cut sedimentary rock blocks intact with a chisel
  - Clump falling for drystack/thatch, build ceilings with supports
  - Better, more survivable gundu fish, enables fish farming
  - Can eat unmelted snow for thirst, with drawbacks
  - Spanish translation


## [2022-05-28][2022-05-28__0.3.7] — 0.3.7 and 0.2.10
  - balance: limited fire spread for multiplayer
  - many airboat improvements, including strafe movement with AUX1
  - Haze overlay warns of extreme heat
  - animal probes to watch the health and age of livestock
  - underwater plants spread and replenish
  - beds and thatch reduce fall damage
  - some valuable nodes have protection in multiplayer
  - beginnings of i18n French translation, early translation support

## [2022-04-16][2022-04-16__0.3.6] — 0.3.6 and 0.2.9
 - Craftable dyes, made from randomly selected plants
 - Wielded_light replaces Illumination mod
 - Iron chests and basic protection for valuable items
 - Trees regrow after a rainy season, not only during rain
 - Transfer water directly between pots and bottles
 - Further cooking pot improvements

## [2022-02-25][2022-02-25__0.3.5b] — 0.3.5b and 0.2.8b
 - flowing water no longer supports player weight in MT 5.5.0
 - /suicide command for players who are stuck
 - added frost overlay to the HUD when you're dangerously cold
 - cooking pot now available, still somewhat WIP
 - many minor bugfixes and improvements

## [2022-01-21][2022-01-21__0.3.5] — 0.3.5 and 0.2.8
  - Blankets to keep warm.
  - Foods can now burn if left cooking too long.
  - Mobs should no longer `stun`-lock when hit repeatedly.
  - Nerf: Increased fall damage.
  - You can use a `Stick` to bar a door, to prevent entry.
  - New animations, thanks to ***MisterE***.
  - Track how many incarnations you’ve had in the `Character` tab.
  - `Glow Worm`s in caves.
  - Nerf: `Stick`s don’t connect with right-angle corners anymore.
  - Nerf: `Mudbrick` turns to clay when broken, in the same manner as bricks lose mortar.
  - Mapgen: `carpathian` is now supported, for a harder game world.
  - No more beach spiders.
  - New food system infrastructure.
  - Experimental biomes, enable under settings to test.
  - Miscellaneous bug-fixes and quality of life improvements.

## [2021-11-01][2021-11-01__0.3.4] — 0.3.4 and 0.2.7
  - New player API: Head tracking, multiplayer skins, male and female characters.
  - Added an introduction screen for new players, and a chat display of the day/season/year with in-universe style.
  - Place snow/ice/water on top of a clay `Water Pot`, and melting will fill it.
  - Buff: Food value of meat.
  - Improved `Torch` stacking by rounding-off burn times.
  - New chat commands: `/set_tempscale`, `/date`.
  - Better indoors check on `Minetest 5.4.x`.

## [2021-09-25][2021-09-25__0.3.3] — 0.3.3 and 0.2.6
  - Plants now grow even when players are away from the area.
  - Fireproof: `Iron Doors` and `Iron Trapdoors`.
  - Ability to mix `Wet Loam` and agricultural soil.
  - Renamed `Stone Chopper` to `Stone Knife` for clarity.
  - More bug fixes and improvements.

## [2021-09-12][2021-09-12__0.3.2] — 0.3.2 and 0.2.5
  - Glassworking! Make windows, and make glass bottles that can hold water, and which stack.
  - Agricultural soils erode down to a soil slab instead of vanishing.
  - Fixed a baking crash.

## [2021-07-18][2021-07-18__0.3.1] — 0.3.1 and 0.2.4
  - `Oil Lamp`s can be switched off with right-click, and lit with `Fire Stick`s.
  - `Oil Lamp`s can hold more oil and now burn oil slower.
  - Various bug-fixes.

## [2021-06-12][2021-06-12__0.3.0] — 0.3.0 and 0.2.3
  - Initial release for `Minetest 5.4.0`, use Exile 0.2 series for 5.3.0 version.
  - New dependency: `naturalslopeslib` provides better hillsides.
  - Ability to throw `Torch`es for light.
  - Settings for `breaktaker`.
  - Ability to display temperature in Celsius, Fahrenheit, or Kelvin. (Using global settings.)
  - Food burns instead of vanishing, thanks to ***ts***.
  - Farming improvements.
  - Fixes for fire and rain.
  - Many fixes for crashes and bugs.

## [2021-02-15][2021-02-15_Mantar] — Now maintained by Mantar

## [2020-12-29][2020-12-29__0.2.2] — 0.2.2
  - New HUD.
  - Character tab.
  - Health effects.
  - Drugs: `Tiku`, `Tang`, `Metastim`.
  - Added `megamorph` surface air-shafts.
  - Fixed clothing infinite tolerance bug.
  - Added `Iron Fittings` for carpentry crafting.
  - Young vs. old `Tangkal` trees.
  - Added `Tiken` woody cane.
  - Cave sediments and mushrooms.
  - Defined `puts_out_fire` group (e.g. mud now puts out fire), and `masonry` group.
  - Created a walkthrough document for players who want one.
  - Added `Loose Wattle`.
  - Added `Mashed Anperla`.
  - Miscellaneous minor fixes and balancing.

## [2020-09-25][2020-09-25__0.2.1] — 0.2.1
  - Fixed exile-letter crash.
  - Fixed saltwater surface sediment crash.
  - Fixed multiplayer health-tab crash.
  - Animals balanced and bug-fixed.
  - Adjusted ice freezing and thawing.

## [2020-08-02][2020-08-02__0.2.0] — 0.2.0

mods/health/hud.lua
----------------------------------------------------------------------
--HUD
----------------------------------------------------------------------

-- Internationalization
local S = HEALTH.S

local hud = {}
local overlaid = {}
local hudupdateseconds = tonumber(minetest.settings:get("exile_hud_update"))

local show_stats = minetest.settings:get_bool("exile_hud_raw_stats") or false

local hud_opacity = minetest.settings:get("exile_hud_icon_transparency") or 127

-- These are color values for the various status levels. They have to be modified
-- per-function below because textures expect one color format and text another.
-- This is a minetest caveat.

-- In texture coloring we simply concat a #.
--		"#"..stat_color

-- In text coloring we concat an 0x and convert the resulting string to a number.
--		tostring("0x"..stat_color)

local stat_fine 	= "FFFFFF"
local stat_slight 	= "FDFF46"
local stat_problem 	= "FF8100"
local stat_major 	= "DF0000"
local stat_extreme	= "8008FF"

local setup_hud = function(player)

	player:hud_set_flags({healthbar = false})
	local playername = player:get_player_name()
	
	local hud_vert_pos 		= -128		-- all HUD icon vertical position
	local hud_extra_y		= -16		-- pixel offset for hot/cold icons
	local hud_text_y		= 32		-- optional text stat offset
	
	local hud_health_x 		= -192
	local hud_hunger_x 		= -128
	local hud_thirst_x 		= -64
	local hud_energy_x 		=  0
	local hud_body_temp_x 	= 64
	local hud_air_temp_x	= 128
	local hud_sick_x		= 192
	
	local icon_scale = {x = 1, y = 1}	-- all HUD icon image scale
	
	local hud_data = {}
	
	hud[playername] = hud_data
	
	hud_data.p_health = player:hud_add({
		hud_elem_type = "image",
		scale = icon_scale,
		offset = {x = hud_health_x, y = hud_vert_pos},
		position = {x = .5, y = 1},
	    text = "hud_health.png^[colorize:#"..stat_fine.."^[opacity:"..hud_opacity
	})
	
	hud_data.p_hunger = player:hud_add({
		hud_elem_type = "image",
		scale = icon_scale,
		offset = {x = hud_hunger_x, y = hud_vert_pos},
		position = {x = .5, y = 1},
	    text = "hud_hunger.png^[colorize:#"..stat_fine.."^[opacity:"..hud_opacity
	})
	
	hud_data.p_thirst = player:hud_add({
		hud_elem_type = "image",
		scale = icon_scale,
		offset = {x = hud_thirst_x, y = hud_vert_pos},
		position = {x = .5, y = 1},
	    text = "hud_thirst.png^[colorize:#"..stat_fine.."^[opacity:"..hud_opacity
	})
	
	hud_data.p_energy = player:hud_add({
		hud_elem_type = "image",
		scale = icon_scale,
		offset = {x = hud_energy_x, y = hud_vert_pos},
		position = {x = .5, y = 1},
	    text = "hud_energy.png^[colorize:#"..stat_fine.."^[opacity:"..hud_opacity
	})
	
	hud_data.p_body_temp = player:hud_add({
		hud_elem_type = "image",
		scale = icon_scale,
		offset = {x = hud_body_temp_x, y = hud_vert_pos},
		position = {x = .5, y = 1},
	    text = "hud_body_temp.png^[colorize:#"..stat_fine.."^[opacity:"..hud_opacity
	})
	
	hud_data.p_body_temp_type = player:hud_add({
		hud_elem_type = "image",
		scale = icon_scale,
		offset = {x = hud_body_temp_x, y = hud_vert_pos + hud_extra_y},
		position = {x = .5, y = 1},
	    text = "hud_temp_normal.png^[opacity:"..hud_opacity -- don't colorize
	})
	
	hud_data.p_air_temp = player:hud_add({
		hud_elem_type = "image",
		scale = icon_scale,
		offset = {x = hud_air_temp_x, y = hud_vert_pos},
		position = {x = .5, y = 1},
	    text = "hud_air_temp.png^[colorize:#"..stat_fine.."^[opacity:"..hud_opacity
	    
	})
	
	hud_data.p_air_temp_type = player:hud_add({
		hud_elem_type = "image",
		scale = icon_scale,
		offset = {x = hud_air_temp_x, y = hud_vert_pos + hud_extra_y},
		position = {x = .5, y = 1},
	    text = "hud_temp_normal.png^[opacity:"..hud_opacity -- don't colorize
	    
	})
	
	hud_data.p_sick = player:hud_add({
		hud_elem_type = "image",
		scale = icon_scale,
		offset = {x = hud_sick_x, y = hud_vert_pos},
		position = {x = .5, y = 1},
	    text = "hud_sick.png^[colorize:#"..stat_fine.."^[opacity:"..hud_opacity
	})
	
	
	hud_data.p_health_text = player:hud_add({
		hud_elem_type = "text",
		offset = {x = hud_health_x, y = hud_vert_pos + hud_text_y},
		number = stat_col,
		position = {x = .5, y = 1},
		text = ""
	})

	hud_data.p_hunger_text = player:hud_add({
		hud_elem_type = "text",
		offset = {x = hud_hunger_x, y = hud_vert_pos + hud_text_y},
		number = stat_col,
		position = {x = .5, y = 1},
		text = ""
	})

	hud_data.p_thirst_text = player:hud_add({
		hud_elem_type = "text",
		offset = {x = hud_thirst_x, y = hud_vert_pos + hud_text_y},
		number = stat_col,
		position = {x = .5, y = 1},
		text = ""
	})

	hud_data.p_energy_text = player:hud_add({
		hud_elem_type = "text",
		offset = {x = hud_energy_x, y = hud_vert_pos + hud_text_y},
		number = stat_col,
		position = {x = .5, y = 1},
		text = ""
	})

	hud_data.p_body_temp_text = player:hud_add({
		hud_elem_type = "text",
		offset = {x = hud_body_temp_x, y = hud_vert_pos + hud_text_y},
		number = stat_col,
		position = {x = .5, y = 1},
		text = ""
	})

	hud_data.p_air_temp_text = player:hud_add({
		hud_elem_type = "text",
		offset = {x = hud_air_temp_x, y = hud_vert_pos + hud_text_y},
		number = stat_col,
		position = {x = .5, y = 1},
		text = ""
	
	})

	hud_data.p_sick_text = player:hud_add({
		hud_elem_type = "text",
		offset = {x = hud_sick_x, y = hud_vert_pos + hud_text_y},
		number = stat_col,
		position = {x = .5, y = 1},
		text = ""
	})

end

minetest.register_on_joinplayer(function(player) setup_hud(player) end)

-- status indicator colors for use in stat display option

local function color(v)
	local stat_col = stat_fine
	if v <= 20 then
		stat_col = stat_extreme
	elseif v <= 40 then
		stat_col = stat_major
	elseif v <= 60 then
		stat_col = stat_problem
	elseif v <= 80 then
		stat_col = stat_slight
	end
	return stat_col
end

local function color_bodytemp(v)
	local stat_col = stat_fine
	local ttype = "hud_temp_normal"
	if v > 47 or v < 27 then
		stat_col = stat_extreme
		if v > 47 then ttype = "hud_temp_hot" end
		if v < 27 then ttype = "hud_temp_cold" end
	elseif v > 43 or v < 32 then
		stat_col = stat_major
		if v > 43 then ttype = "hud_temp_hot" end
		if v < 32 then ttype = "hud_temp_cold" end
	elseif v > 38 or v < 37 then
		stat_col = stat_problem
		if v > 38 then ttype = "hud_temp_hot" end
		if v < 37 then ttype = "hud_temp_cold" end
	end
	return stat_col, ttype
end

local function color_envirotemp(v, meta)
	--make sure matches actual values used!
	local comfort_low = meta:get_int("clothing_temp_min")
	local comfort_high = meta:get_int("clothing_temp_max")
	local stress_low = comfort_low - 10
	local stress_high = comfort_high + 10
	local danger_low = stress_low - 40
	local danger_high = stress_high +40
	local overlay

	local stat_col = stat_fine
	local ttype = "hud_temp_normal"

	if v > danger_high or v < danger_low then
		icon_col = "extreme"
		stat_col = stat_extreme
		if v > danger_high then ttype = "hud_temp_hot" end
		if v < danger_low then ttype = "hud_temp_cold" end
	elseif v > stress_high or v < stress_low then
		stat_col = stat_major
		if v > stress_high then ttype = "hud_temp_hot" end
		if v < stress_low then ttype = "hud_temp_cold" end
	elseif v > comfort_high or v < comfort_low then
		stat_col = stat_slight
		if v > comfort_high then ttype = "hud_temp_hot" end
		if v < comfort_low then ttype = "hud_temp_cold" end
	end
	if v < stress_low then
	   overlay = "weather_hud_frost.png"
	end
	if v > stress_high then
	   overlay = "weather_hud_heat.png"
	end

	return stat_col, ttype, overlay
end


local function health(player, hud_data)
	local v = player:get_hp()
	v = (v/20)*100
	local stat_col = color(v)
	local t = v .." %"
	local hud = hud_data.p_health
	player:hud_change(hud, "text", "hud_health.png^[colorize:#"..stat_col.."^[opacity:"..hud_opacity)
	local hud2 = hud_data.p_health_text
	player:hud_change(hud2, "number", tonumber("0x"..stat_col))
	if show_stats then
		player:hud_change(hud2, "text", t)
	else
		player:hud_change(hud2, "text", "")
	end
end

local function energy(player, hud_data, meta)
	local v = meta:get_int("energy")
	v = (v/1000)*100
	local stat_col = color(v)
	local t = v .." %"
	local hud = hud_data.p_energy
	player:hud_change(hud, "text", "hud_energy.png^[colorize:#"..stat_col.."^[opacity:"..hud_opacity)
	local hud2 = hud_data.p_energy_text
	player:hud_change(hud2, "number", tonumber("0x"..stat_col))
	if show_stats then
		player:hud_change(hud2, "text", t)
	else
		player:hud_change(hud2, "text", "")
	end
end

local function thirst(player, hud_data, meta)
	local v = meta:get_int("thirst")
	v = (v/100)*100
	local t = v .." %"
	local stat_col = color(v)
	local hud =  hud_data.p_thirst
	player:hud_change(hud, "text", "hud_thirst.png^[colorize:#"..stat_col.."^[opacity:"..hud_opacity)
	local hud2 = hud_data.p_thirst_text
	player:hud_change(hud2, "number", tonumber("0x"..stat_col))
	if show_stats then
		player:hud_change(hud2, "text", t)
	else
		player:hud_change(hud2, "text", "")
	end
end

local function hunger(player,  hud_data, meta)
	local v = meta:get_int("hunger")
	v = (v/1000)*100
	local t = v .." %"
	local stat_col = color(v)
	local hud =  hud_data.p_hunger
	player:hud_change(hud, "text", "hud_hunger.png^[colorize:#"..stat_col.."^[opacity:"..hud_opacity)
	local hud2 = hud_data.p_hunger_text
	player:hud_change(hud2, "number", tonumber("0x"..stat_col))
	if show_stats then
		player:hud_change(hud2, "text", t)
	else
		player:hud_change(hud2, "text", "")
	end
end


local function temp(player, hud_data, meta)
	local v = meta:get_int("temperature")
	local stat_col, ttype, overlay = color_bodytemp(v)
	local t = climate.get_temp_string(v, meta)
	local hud =  hud_data.p_body_temp
	local hud2 = hud_data.p_body_temp_type
	player:hud_change(hud, "text", "hud_body_temp.png^[colorize:#"..stat_col.."^[opacity:"..hud_opacity)
	player:hud_change(hud2, "text", ttype..".png^[opacity:"..hud_opacity) -- don't colorize)
	local hud2 = hud_data.p_body_temp_text
	player:hud_change(hud2, "number", tonumber("0x"..stat_col))
	if show_stats then
		player:hud_change(hud2, "text", t)
	else
		player:hud_change(hud2, "text", "")
	end
end

local function do_overlay(player, pname, pos, overlay)
   local handle = player:hud_add({
	 name = overlay,
	 hud_elem_type = "image",
	 position = {x = 0, y = 0},
	 alignment = {x = 1, y = 1},
	 scale = { x = -100, y = -100},
	 z_index = hud.z_index,
	 text = overlay,
	 offset = {x = 0, y = 0}
   })
   overlaid[pname] = handle
end


local function enviro_temp(player, hud_data, meta)
	local pname = player:get_player_name()
	local player_pos = player:get_pos()
	player_pos.y = player_pos.y + 0.6 --adjust to body height
	local v = math.floor(climate.get_point_temp(player_pos))
	local stat_col, ttype, overlay = color_envirotemp(v, meta)
	if overlay then
	   if not overlaid[pname] then
	      do_overlay(player, pname, player_pos, overlay)
	   elseif player:hud_get(overlaid[pname]) and
	      ( overlay ~= player:hud_get(overlaid[pname]).name ) then
	      -- direct transition from one overlay to another
	      player:hud_remove(overlaid[pname])
	      do_overlay(player, pname, player_pos, overlay)
	   end
	elseif overlaid[pname] then -- remove overlay
	   player:hud_remove(overlaid[pname])
	   overlaid[pname] = nil
	end
	local t = climate.get_temp_string(v, meta)
	local newhud = hud_data.p_air_temp
	local newhud2 = hud_data.p_air_temp_type
	player:hud_change(newhud, "text", "hud_air_temp.png^[colorize:#"..stat_col.."^[opacity:"..hud_opacity)
	player:hud_change(newhud2, "text", ttype..".png^[opacity:"..hud_opacity) -- don't colorize)
	local hud2 = hud_data.p_air_temp_text
	player:hud_change(hud2, "number", tonumber("0x"..stat_col))
	if show_stats then
		player:hud_change(hud2, "text", t)
	else
		player:hud_change(hud2, "text", "")
	end
end

local function effects(player, hud_data, meta)
	local stat_col = stat_fine
	local v = meta:get_int("effects_num")
	local t = "x"..v
	if v > 0 then
		stat_col = stat_slight
	elseif v > 1 then
		stat_col = stat_problem
	elseif v > 2 then
		stat_col = stat_major
	elseif v > 3 then
		stat_col = stat_extreme
	end
	local hud = hud_data.p_sick
	player:hud_change(hud, "text", "hud_sick.png^[colorize:#"..stat_col.."^[opacity:"..hud_opacity)
	local hud2 = hud_data.p_sick_text
	player:hud_change(hud2, "number", tonumber("0x"..stat_col))
	if show_stats then
		player:hud_change(hud2, "text", t)
	else
		player:hud_change(hud2, "text", "")
	end
end

local timer = 0

minetest.register_globalstep(function(dtime)
  timer = timer + dtime

  if timer > hudupdateseconds then
   for _0, player in ipairs(minetest.get_connected_players()) do

		local name = player:get_player_name()
		local meta = player:get_meta()
		local hud_data = hud[name]
		if not hud_data then
			return
		end

		health(player, hud_data)
		energy(player, hud_data, meta)
		thirst(player, hud_data, meta)
		hunger(player, hud_data, meta)
		temp(player, hud_data, meta)
		enviro_temp(player, hud_data, meta)
		effects(player, hud_data, meta)

   end
   timer = 0
   return nil
  end
end)

minetest.register_chatcommand("show_stats", {
	params = "help",
    description = "Enable or disable stats showing below icons.",
    func = function(name, param)
		local var = minetest.settings:get_bool("exile_hud_raw_stats")
		if param == "help" then
			local wlist = "/show_stats:\n"..
			"Toggle stats showing below icons."
			return false, wlist
		else	
			if var then
				minetest.settings:set_bool("exile_hud_raw_stats", false)
				show_stats = false
			else
				minetest.settings:set_bool("exile_hud_raw_stats", true)
				show_stats = true
			end
		end   
	end
})

minetest.register_chatcommand("icon_transparency", {
    params = "<int>",
    description = "Set stat icon transparency between 0 and 255 or default (127)",
    func = function(name, param)
		if param == "" or param == "help" then
			local wlist = "/show_stats:\n"..
			"Set stat icon transparency between 0 and 255.\n"..
			"Valid value is an integer between 0 and 255 or default.\n"
			return false, wlist
		end
		if param == "default" then
			minetest.settings:set("exile_hud_icon_transparency", 127)
			hud_opacity = 127
		else
			local num = tonumber(param)
			if type(num) == "number" then
				if num < 0 then minetest.settings:set("exile_hud_icon_transparency", 0)
				hud_opacity = 0
				return false, "Icon transparency set to 0"
				elseif num > 255 then minetest.settings:set("exile_hud_icon_transparency", 255)
				hud_opacity = 255
				return false, "Icon transparency set to 255"
				else minetest.settings:set("exile_hud_icon_transparency", math.floor(num))
				hud_opacity = num
				return false, "Icon transparency set to "..math.floor(num)
				end
			else
				return false, "Invalid value. Please use a whole number between 0 and 255"
			end
		end	
	end
})
